implementations/rust/ockam/ockam_api/src/authenticator/direct/direct_authenticator.rs

```diff
@@ -0,0 +1,155 @@
+use either::Either;
+use std::collections::{BTreeMap, HashMap};
+
+use ockam::identity::utils::now;
+use ockam::identity::AttributesEntry;
+use ockam::identity::Identifier;
+use ockam_core::compat::sync::Arc;
+use ockam_core::Result;
+
+use crate::authenticator::{
+    AuthorityMember, AuthorityMembersRepository, EnrollerAccessControlChecks,
+};
+
+/// Identity attribute key that indicates the role of the subject
+pub const OCKAM_ROLE_ATTRIBUTE_KEY: &str = "ockam-role";
+
+/// Identity attribute value that indicates the enroller role of the subject
+/// the corresponding key is [`OCKAM_ROLE_ATTRIBUTE_KEY`]
+pub const OCKAM_ROLE_ATTRIBUTE_ENROLLER_VALUE: &str = "enroller";
+
+pub struct DirectAuthenticatorErrorString(pub String);
+
+pub struct DirectAuthenticator {
+    members: Arc<dyn AuthorityMembersRepository>,
+}
+
+impl DirectAuthenticator {
+    pub fn new(members: Arc<dyn AuthorityMembersRepository>) -> Self {
+        Self { members }
+    }
+
+    pub async fn add_member(
+        &self,
+        enroller: &Identifier,
+        id: &Identifier,
+        attrs: &BTreeMap<String, String>,
+    ) -> Result<Either<(), DirectAuthenticatorErrorString>> {
```

Comment:
Maybe a type alias `type DirectAuthenticatorResult<a> = Either<a, DirectAuthenticatorError>` would make the function signatures slightly better?